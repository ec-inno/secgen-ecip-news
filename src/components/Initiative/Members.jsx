import React from 'react';
import has from 'lodash/has';

import getCurrentLanguage from '../../utils/getCurrentLanguage';
import getDefaultLanguage from '../../utils/getDefaultLanguage';

const Members = ({ initiativeData }) => {
  const language = getCurrentLanguage(location) || getDefaultLanguage();
  const translation = require(`../../../translations/initiative/${language}.json`);

  // [REPRESENTATIVE, SUBSTITUTE, MEMBER, LEGAL_ENTITY, OTHER, DPO ]
  const people = has(initiativeData, 'members')
    ? initiativeData.members
        .map(p => {
          const fullName = p.activityPeriod
            ? `${fullName} (${p.activityPeriod})`
            : fullName;
          return {
            fullName,
            ...p,
          };
        })
        .filter(p => !p.privacyApplied)
    : [];
  const reps = people.filter(p => p.type === 'REPRESENTATIVE');
  const subs = people.filter(p => p.type === 'SUBSTITUTE');
  const members = people.filter(p => p.type === 'MEMBER');
  const legalEs = people.filter(p => p.type === 'LEGAL_ENTITY');
  const others = people.filter(p => p.type === 'OTHER');
  const dpos = people.filter(p => p.type === 'DPO');

  return (
    <>
      {reps || subs || members || legalEs || others || dpos ? (
        <>
          <h2 className="ecl-u-type-heading-2">{translation.organisers}</h2>
          <ul className="ecl-u-type-paragraph">
            {legalEs && legalEs.length > 0 ? (
              <li key="legalEntity">
                Legal entities
                {': '}
                {legalEs.map(m => m.fullName).join(', ')}
              </li>
            ) : (
              ''
            )}
            {reps && reps.length > 0
              ? reps.map((rep, key) => (
                  <li key={`rep-${key}`}>
                    {translation.representative}
                    {': '}
                    {rep.fullName}
                    {rep.email ? ` - ${rep.email}` : ''}
                  </li>
                ))
              : ''}
            {subs && subs.length > 0
              ? subs.map((sub, key) => (
                  <li key={`sub-${key}`}>
                    {translation.substitute}
                    {': '}
                    {sub.fullName}
                    {sub.email ? ` - ${sub.email}` : ''}
                  </li>
                ))
              : ''}
            {members && members.length > 0 ? (
              <li key="members">
                {translation.members}
                {': '}
                {members.map(m => m.fullName).join(', ')}
              </li>
            ) : (
              ''
            )}
            {others && others.length > 0 ? (
              <li key="others">
                Others
                {': '}
                {others.map(m => m.fullName).join(', ')}
              </li>
            ) : (
              ''
            )}
            {dpos && dpos.length > 0 ? (
              <li key="dpos">
                DPO
                {': '}
                {dpos.map(m => m.fullName).join(', ')}
              </li>
            ) : (
              ''
            )}
          </ul>
        </>
      ) : (
        ''
      )}
    </>
  );
};

export default Members;
