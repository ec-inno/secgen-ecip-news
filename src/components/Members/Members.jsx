import React from 'react';
import PropTypes from 'prop-types';
import { useTranslation } from 'react-i18next';

const Members = ({ members }) => {
  if (!members || !Array.isArray(members)) return '';

  // [REPRESENTATIVE, SUBSTITUTE, MEMBER, LEGAL_ENTITY, OTHER, DPO ]
  const people = members
    ? members
        .filter(person => person.fullName)
        .filter(person => person.privacyApplied === false)
        .map(person => {
          const { activityPeriod, fullName: name } = person;

          // Include information about time of activity if present.
          const fullName = activityPeriod
            ? `${name} (${activityPeriod})`
            : name;

          return {
            ...person,
            fullName,
          };
        })
    : [];

  if (people.length === 0) return '';

  const reps = people.filter(p => p.type === 'REPRESENTATIVE');
  const subs = people.filter(p => p.type === 'SUBSTITUTE');
  const mem = people.filter(p => p.type === 'MEMBER');
  const legalEs = people.filter(p => p.type === 'LEGAL_ENTITY');
  const others = people.filter(p => p.type === 'OTHER');
  const dpos = people.filter(p => p.type === 'DPO');

  if (!reps && !subs && !mem && !legalEs && !others && !dpos) return '';

  const { t } = useTranslation();

  return (
    <>
      <h2 className="ecl-u-type-heading-2">{t('Organisers')}</h2>
      <ul className="ecl-u-type-paragraph">
        {legalEs && legalEs.length > 0 ? (
          <li key="legalEntity">
            {t('Legal entities')}
            {': '}
            {legalEs.map(m => m.fullName).join(', ')}
          </li>
        ) : (
          ''
        )}
        {reps && reps.length > 0
          ? reps.map((rep, key) => (
              <li key={`rep-${key}`}>
                {t('Representative')}
                {': '}
                {rep.fullName}
                {rep.email ? ` - ${rep.email}` : ''}
              </li>
            ))
          : ''}
        {subs && subs.length > 0
          ? subs.map((sub, key) => (
              <li key={`sub-${key}`}>
                {t('Substitude')}
                {': '}
                {sub.fullName}
                {sub.email ? ` - ${sub.email}` : ''}
              </li>
            ))
          : ''}
        {mem && mem.length > 0 ? (
          <li key="members">
            {t('Members')}
            {': '}
            {mem.map(m => m.fullName).join(', ')}
          </li>
        ) : (
          ''
        )}
        {others && others.length > 0 ? (
          <li key="others">
            {t('Others')}
            {': '}
            {others.map(m => m.fullName).join(', ')}
          </li>
        ) : (
          ''
        )}
        {dpos && dpos.length > 0 ? (
          <li key="dpos">
            {t('DPO')}
            {': '}
            {dpos.map(m => m.fullName).join(', ')}
          </li>
        ) : (
          ''
        )}
      </ul>
    </>
  );
};

Members.propTypes = {
  members: PropTypes.arrayOf(
    PropTypes.shape({
      type: PropTypes.string,
      fullName: PropTypes.string,
      privacyApplied: PropTypes.bool,
    })
  ),
};

export default Members;
