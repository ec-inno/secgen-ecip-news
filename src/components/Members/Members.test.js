import React from 'react';
import renderer from 'react-test-renderer';
import Members from './Members';

describe('Members', () => {
  it('renders an empty string on missing input', () => {
    const tree = renderer.create(<Members />).toJSON();
    expect(tree).toBe('');
  });

  it('respects privacyApplied flag', () => {
    const props = {
      members: [
        {
          type: 'MEMBER',
          fullName: 'John DOE',
          privacyApplied: false,
        },
        {
          type: 'MEMBER',
          fullName: 'Jane DOE',
          privacyApplied: true,
        },
      ],
    };

    const tree = renderer.create(<Members {...props} />).toJSON();
    const list = tree[1].children.filter(m => m);

    expect(tree[1].type).toBe('ul');
    expect(list.length).toBe(1);
    expect(tree).toMatchSnapshot();
  });

  it('Handles members of the following types: REPRESENTATIVE, SUBSTITUTE, MEMBER, LEGAL_ENTITY, OTHER, DPO', () => {
    const props = {
      members: [
        {
          type: 'REPRESENTATIVE',
          fullName: 'John REPRESENTATIVE',
          privacyApplied: false,
        },
        {
          type: 'SUBSTITUTE',
          fullName: 'John SUBSTITUDE',
          privacyApplied: false,
        },
        {
          type: 'MEMBER',
          fullName: 'John MEMBER',
          privacyApplied: false,
        },
        {
          type: 'LEGAL_ENTITY',
          fullName: 'John LEGAL_ENTITY',
          privacyApplied: false,
        },
        {
          type: 'OTHER',
          fullName: 'John OTHER',
          privacyApplied: false,
        },
        {
          type: 'DPO',
          fullName: 'John DPO',
          privacyApplied: false,
        },
      ],
    };

    const tree = renderer.create(<Members {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
