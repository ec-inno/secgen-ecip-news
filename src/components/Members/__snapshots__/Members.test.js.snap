// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Members Handles members of the following types: REPRESENTATIVE, SUBSTITUTE, MEMBER, LEGAL_ENTITY, OTHER, DPO 1`] = `
Array [
  <h2
    className="ecl-u-type-heading-2"
  >
    Organisers
  </h2>,
  <ul
    className="ecl-u-type-paragraph"
  >
    <li>
      Legal entities
      : 
      John LEGAL_ENTITY
    </li>
    <li>
      Representative
      : 
      John REPRESENTATIVE
      
    </li>
    <li>
      Substitude
      : 
      John SUBSTITUDE
      
    </li>
    <li>
      Members
      : 
      John MEMBER
    </li>
    <li>
      Others
      : 
      John OTHER
    </li>
    <li>
      DPO
      : 
      John DPO
    </li>
  </ul>,
]
`;

exports[`Members respects privacyApplied flag 1`] = `
Array [
  <h2
    className="ecl-u-type-heading-2"
  >
    Organisers
  </h2>,
  <ul
    className="ecl-u-type-paragraph"
  >
    
    
    
    <li>
      Members
      : 
      John DOE
    </li>
    
    
  </ul>,
]
`;
